lib/hanami/action/mime.rb

```diff
@@ -565,33 +560,41 @@ def content_type_with_charset
       # @see https://github.com/rack/rack/pull/659
       # @see https://github.com/hanami/controller/issues/59
       # @see https://github.com/hanami/controller/issues/104
+      # @see https://github.com/hanami/controller/issues/275
       def best_q_match(q_value_header, available_mimes)
-        values = ::Rack::Utils.q_values(q_value_header)
-        values = values.map do |req_mime, quality|
-          if req_mime == DEFAULT_ACCEPT
-            # See https://github.com/hanami/controller/issues/167
-            match = default_content_type
-          else
-            match = available_mimes.find { |am| ::Rack::Mime.match?(am, req_mime) }
-          end
+        values = []
+        ::Rack::Utils.q_values(q_value_header).each_with_index do |(req_mime, quality), index|
+          match = available_mimes.find { |am| ::Rack::Mime.match?(am, req_mime) }
           next unless match
-          [match, quality]
-        end.compact
-
-        if Hanami::Utils.jruby?
-          # See https://github.com/hanami/controller/issues/59
-          # See https://github.com/hanami/controller/issues/104
-          values.reverse!
-        else
-          # See https://github.com/jruby/jruby/issues/3004
-          values.sort!
+          values << Specification.new(req_mime, quality, index, match)
         end
 
-        value = values.sort_by do |match, quality|
-          (match.split('/'.freeze, 2).count('*'.freeze) * -10) + quality
-        end.last
+        value = values.max
+        value.format if value
+      end
+
+      class Specification
+        include Comparable
 
-        value.first if value
+        attr_reader :quality, :index, :mime, :format, :priority
+        def initialize(mime, quality, index, format = mime)
+          @quality, @index, @format = quality, index, format
+          calculate_priority(mime)
+        end
+
+        def <=>(other)
+          return priority <=> other.priority unless priority == other.priority
```

Comment:
Yep.

Since priority already factors the quality of the mime type, there's no need to compare based into it. TBH I starting by comparing it, ran some tests, got weird results and then realized that.

Oh, and thanks for the review @TiteiKo =)